crates/storage/trie/node/branch.rs

```diff
@@ -0,0 +1,665 @@
+                let choice_index = Nibble::try_from(choice_index as u8).unwrap();
+                let child_node = db
+                    .get_node(*child_ref)?
+                    .expect("inconsistent internal tree structure");
+
+                match child_node {
+                    // Replace the child node  with an extension node leading to it
+                    // The extension node will then replace self if self has no value
+                    Node::Branch(_) => {
+                        *child_ref = db.insert_node(
+                            ExtensionNode::new(
+                                NibbleVec::from_single(choice_index, path_offset % 2 != 0),
+                                *child_ref,
+                            )
+                            .into(),
+                        )?;
+                    }
+                    // Replace self with the child extension node, updating its path in the process
+                    Node::Extension(mut extension_node) => {
+                        debug_assert!(self.path.is_empty()); // Sanity check
+                        extension_node.prefix.prepend(choice_index);
+                        // Return node here so we don't have to update it in the DB and then fetch it
+                        return Ok((Some(extension_node.into()), value));
+                    }
+                    _ => {}
+                }
+
+                Some(child_ref)
+            }
+            _ => None,
+        };
+
+        let new_node = match (child_ref, !self.path.is_empty()) {
+            // If this node still has a child and value return the updated node
+            (Some(_), true) => Some(self.into()),
+            // If this node still has a value but no longer has children, convert it into a leaf node
+            (None, true) => Some(LeafNode::new(self.path, self.value).into()),
+            // If this node doesn't have a value, replace it with its child node
+            (Some(x), false) => Some(
+                db.get_node(*x)?
+                    .expect("inconsistent internal tree structure"),
+            ),
+            // Return this node
+            (None, false) => Some(self.into()),
+        };
+
+        Ok((new_node, value))
+    }
+
+    /// Computes the node's hash given the offset in the path traversed before reaching this node
+    pub fn compute_hash(&self, db: &TrieDB, path_offset: usize) -> Result<NodeHashRef, StoreError> {
+        if let Some(hash) = self.hash.extract_ref() {
+            return Ok(hash);
+        };
+        let hash_choice = |node_ref: NodeRef| -> Result<DelimitedHash, StoreError> {
+            if node_ref.is_valid() {
+                let child_node = db
+                    .get_node(node_ref)?
+                    .expect("inconsistent internal tree structure");
+
+                let mut target = Output::default();
+                let target_len = match child_node.compute_hash(db, path_offset + 1)? {
```

Comment:
I think this is probably what "fixes" the fact that we're not addressing by hash. The unfortunate thing about this is that it means loading potentially the whole trie in the process.

Reply:
This is not really a problem of not storing by hash as the nodes themselves can store their own hash and return it upon request